src/liballoc/collections/btree/node.rs

```diff
@@ -489,32 +527,42 @@ impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
         // SAFETY: The values of a node must always be initialized up to length.
         unsafe {
             slice::from_raw_parts_mut(
-                MaybeUninit::first_ptr_mut(&mut (*self.as_leaf_mut()).vals),
+                MaybeUninit::first_ptr_mut(&mut self.as_leaf_mut().vals),
                 self.len(),
             )
         }
     }
 
-    fn into_slices_mut(mut self) -> (&'a mut [K], &'a mut [V]) {
-        // We cannot use the getters here, because calling the second one
-        // invalidates the reference returned by the first.
```

Comment:
Might still be worth explaining why we do not call `into_key_mut_at` and then `into_val_mut_at`. That was the old implementation and I fear someone will revert it back to that if there is no comment explaining this.

Reply:
Indeed, I began thinking all aliasing was resolved in the `Mut` case but it isn't.

Reply:
I was starting to doubt this point. I don't mean because `into_key_mut_at` and `into_val_mut_at`don't exist. One can argue they should exist, because each of the invocations of `into_kv_mut`in map.rs wants only the key or value, and I don't believe that this is all optimized away to a single array access. So I added them for performance alone.

I do mean because the borrowing rules I'm beginning to understand again, already guarantee that you can't invoke both. It says "cannot move out of `self` because it is borrowed" to this:

```
        let key = unsafe { self.reborrow_mut().into_key_mut_at(idx) };
        let val = unsafe { self.into_val_mut_at(idx) };
```

The slices were references too, so why didn't the borrow checker complain much earlier? I think it was because the old code didn't try to reborrow but bluntly copied the pointers:

```
   fn into_slices_mut(self) -> (&'a mut [K], &'a mut [V]) {
        let k = unsafe { ptr::read(&self) };
        (k.into_key_slice_mut(), self.into_val_slice_mut())
    }
```

Not that this is a reason to omit the comment, but rather to elaborate it.

Reply:
> I think it was because the old code didn't try to reborrow but bluntly copied the pointers:

Indeed, it just duplicated `self` to be able to call these two methods.